iroha_p2p/src/peer.rs

```diff
@@ -29,422 +27,701 @@ use ursa::{
-                    return Err(Error::Keys);
-                }
-            },
+            Some(cipher) => decrypt_easy(cipher, &data)?,
         };
 
-        let pub_key: Result<iroha_crypto::PublicKey, _> = Decode::decode(&mut data.as_slice());
-        match pub_key {
-            Ok(pub_key) => {
-                self.id.public_key = pub_key;
-                self.state = State::Ready;
-                Ok(())
-            }
+        let pub_key = match Decode::decode(&mut data.as_slice()) {
+            Ok(pub_key) => pub_key,
             Err(e) => {
                 warn!(%e, "Unexpected error creating encryptor!");
-                self.state = State::Error;
-                Err(Error::Keys)
+                return Err(Error::Keys);
             }
+        };
+        self.id.public_key = pub_key;
+        self.state = ToState::Ready;
+        Ok(())
+    }
+
+    async fn handshake(&mut self) -> Result<(), Error> {
+        debug!(?self.state, addr = %self.id.address, "Attempting handshake");
+        match self.state {
+            ToState::Connected => self.read_hello().await?,
+            ToState::SendKey => self.send_public_key().await?,
+            ToState::GetKey => self.read_public_key().await?,
+            ToState::Ready => warn!("Not doing handshake, already ready."),
         }
+        Ok(())
     }
 
-    /// Creates shared key from two public keys - our and their,
-    /// and creates and encryptor from that key.
-    fn derive_shared_key(&mut self, public_key: &PublicKey) -> Result<(), Error> {
-        let dh = K::new();
-        let shared = match dh.compute_shared_secret(&self.secret_key, public_key) {
-            Ok(key) => key,
-            Err(e) => {
-                warn!(%e, "Error creating shared secret!");
-                return Err(Error::Keys);
-            }
-        };
-        debug!("Derived shared key: {:?}", &shared.0);
-        match self.new_encryptor(shared.0.as_slice()) {
-            Ok(encryptor) => {
-                self.cipher = Some(encryptor);
-                Ok(())
-            }
-            Err(e) => {
-                warn!(%e, "Unexpected error creating encryptor!");
-                Err(Error::Keys)
-            }
+    async fn into_peer(self) -> Addr<Peer<T, K, E>> {
+        let (r, w) = self.stream.into_split();
+        let (sender, receiver) = oneshot::channel();
+        let peer = Peer::new(
+            self.id,
+            w,
+            self.cipher.unwrap(),
+            self.network,
+            false,
+            sender,
+        );
+        let stream = Peer::<T, K, E>::stream(r, receiver);
+
+        let peer = peer.preinit();
+        let addr = peer.address.clone();
+
+        // spawning task
+        tokio::spawn(async move {
+            stream.for_each(|it| async { addr.do_send(it).await }).await;
+        });
+
+        peer.start().await
+    }
+}
+
+#[derive(iroha_actor::Message)]
+struct ContinueHandshake;
```

Comment:
This message means "start connection" also, not continue handshakes only.

Reply:
I guess it can be renamed to just `Handshake`

Reply:
But there is a connection attempt also.

Reply:
Hmm, not sure how to name it then

Reply:
Just make a simple loop and remove this message for good :)

Reply:
The point of it is to let peers be stopped by network even then they are handshaking